MachObfuscator/Mach/ObjC/MachStrings.swift

```diff
@@ -0,0 +1,52 @@
+//
+//  Mach+Strings.swift
+//  MachObfuscator
+//
+
+import Foundation
+
+protocol ContainedInData {
+    // associatedtype ValueType
+    init(value: String, range: Range<Int>)
+    var value: String { get }
+    var range: Range<Int> { get }
+}
+
+class StringInData: ContainedInData, CustomStringConvertible {
```

Comment:
Looks like `ContainedInData` is almost never used.

Do you think we could just remove `ContainedInData` and use `StringInData` everywhere? This would simplify a lot of generic methods.

Do you think we could rename `StringInData` as `RangedString`? I think `StringInData` does not fully express the role of this class.

Why don't making this a struct? I can see this is a class to allow inheritance. But maybe `MangledObjcClassNameInData` isn't worth it? `isSwiftName` is used in just one place, so maybe you could move `MangledObjcClassNameInData.isSwiftName` to `private extension RangedString` in `Mach+classNames.swift`?

Reply:
`StringInData` is a base for more classes that decode different kinds of mangled Swift identifiers. That is why it is a class. Having separate classes increases code readability: it is explicit what format of the string to expect in given case (eg. is it mangled or plain name), where private extension does not provide this information. For example, you should not call `isSwiftName` on selector name - it does not make sense.

Reply:
Naming things is hard:)

`RangedString` for me means `String` and `Range`, which is not all I want to tell by implementing this protocol.

`ContainedInData` is supposed to mean that this object was loaded from mach-o image and lives at a specific place in it (implemented using `Range` - but it is an implementation detail). 

`ContainedInData` is not a perfect name either. Maybe `MetadataValue` or `MachoMetadataObject`?

Reply:
I cleaned up usage of `ContainedInData` (`getCString` should return string, not any object). Have a look.

Reply:
So we have 2 sub-discussions here :)

1) `StringInData` being a class.
2) `StringInData` naming.

Let's continue the discussion in two paralel threads.

## Regarding 1)

I can understand your motives... more types are coming here :) The idea about `MangledObjcClassNameInData.isSwiftName` is ok for me. I just don't like the inheritance and I always try to avoid inheritance if only possible. Don't ask me about the reasons, because you probably know all of them :D I think we can still drop the classes here, and replace them with structs keeping you yet-to-come classes still functional and small.

Take a look at following code:
```
struct StringInData: ContainedInData, CustomStringConvertible {
    let value: String
    let range: Range<Int>
    // having init for free ;)
}

struct MangledObjcClassNameInData: ContainedInData {
    let value: String
    let range: Range<Int>
}

extension MangledObjcClassNameInData
    var isSwiftName: Bool {
        return value.starts(with: "_Tt")
    }
}
```
All the functions with `where R: StringInData` could then get `where R: ContainedInData, R.ValueType == String`.
What do you think about this?

## Regarding 2)

> `Range` - but it is an implementation detail

It is not an implementation detail – it is not private and the `range` property is used externally by other functions. But I can get your point – the `ContainedInData` is not about `range`, it is about possibility of being found in the source data. I think nor `ContainedInData`, neither `StringInData` expresses that point. It still requires a cognitive load to understand the role of this class.

I still think the `RangedData` is a good name, but I also have some alternatives :) `ValueReference`, `MachOValueReference`, `DataReference`, `Subdata`, `MachOSubdata`, `DataToken`, `MachODataToken`, `AddressedValue`.

Reply:
Regarding 1: I changed the code the way you suggest. 

I am not a very big fan of protocol inheritance, because I have already fought with its limitations. For example, if a protocol has an associated type, you cannot define variable or property in class using this protocol (you can use generics, but this makes things more complicated).

Some implementation inheritance would be useful. As you can see, I had to duplicate precondition that check invariants for length.

Regarding 2: Thanks for the proposals:) I am still thinking how to name this thing.

Reply:
I like the `MachO` prefix (these classes are supposed to be used with data originating from MachO files, not with any `Data` even though their API does not enforce this). My suggestion:

- `ContainedInData` -> `MachOValue`
- `PlainStringInData` (previously `StringInData`) -> `MachOPlainString`
- `MangledObjcClassNameInData` -> `MachOMangledObjcClassName`

What do you think?

Reply:
Regarding 1: I have heard about those problems, just never really had in real live. Do you think those problems are going to hit us in your next PR? Let's keep those structs and think about solving those problems when they occur. I just tried to stay away from reference semantics all across the MachObfuscator. Precondition duplication is really just a kind of trade-off.

Regarding 2: What do you think about merging this PR, and eventually fixing the name in the next PR?

Reply:
Regarding 1: For example, if you wanted to create (a quite natural and elegant) protocol specific for string: `protocol StringContainedInData: ContainedInData where ValueType == String {}` then you cannot use it as a property type, array element or return value because you would get compiler error (in Xcode 10.2.1) `Protocol 'StringContainedInData' can only be used as a generic constraint because it has Self or associated type requirements` even though associatedtype was bound explicitly. One solution for this is type erasure, but then you have to define another set of protocols which adds complexity (I used this approach in ObjcSymbols vs ObjcSymbols+Impl).

Another no so nice situation is when you want to extend the protocol to conform with another protocol (which is simple when you add conformance to structs/classes). It is not possible directly, it can be done the way I implemented `CustomStringConvertible` for `ContainedInData`. I had to think for a while to understand what it does and class conforming to `ContainedInData` must also explicitly declare `CustomStringConvertible` to use the default `description` implementation.

Regarding 2: I would like to set the names in this PR - it will be cleaner. We just have to agree what the names should be :)

Reply:
Regarding 2: sorry for too-early merge :(